File: src/windows_worker/trigger.py
from __future__ import annotations

import re
from dataclasses import dataclass

from windows_worker.config import WechatListenSettings
from windows_worker.wechat_adapter import WechatMessage


@dataclass(frozen=True)
class TriggerMatch:
    group_id: str
    group_name: str | None
    sender_id: str
    sender_name: str | None
    trigger_content: str
    trigger_symbol: str
    timestamp: int


class WindowsTriggerMatcher:
    def __init__(self, settings: WechatListenSettings):
        self.settings = settings
        self._regexes = (
            [re.compile(pattern) for pattern in settings.triggers]
            if settings.match_mode == "regex"
            else []
        )

    def match(self, message: WechatMessage) -> TriggerMatch | None:
        if not self._group_allowed(message.group_id, message.group_name):
            return None
        if message.sender_id in self.settings.blacklist_users:
            return None
        if message.is_self and self.settings.ignore_self:
            return None
        if message.type not in self.settings.content_types:
            return None
        symbol = self._match_symbol(message.content)
        if symbol is None:
            return None
        return TriggerMatch(
            group_id=message.group_id,
            group_name=message.group_name,
            sender_id=message.sender_id,
            sender_name=message.sender_name,
            trigger_content=message.content,
            trigger_symbol=symbol,
            timestamp=message.timestamp,
        )

    def _group_allowed(self, group_id: str, group_name: str | None) -> bool:
        if not self.settings.whitelist_groups:
            return True
        name = group_name or ""
        return any(
            item == group_id or (name and item in name)
            for item in self.settings.whitelist_groups
        )

    def _match_symbol(self, content: str) -> str | None:
        if self.settings.match_mode == "prefix":
            return next(
                (symbol for symbol in self.settings.triggers if content.startswith(symbol)),
                None,
            )
        if self.settings.match_mode == "contains":
            return next((symbol for symbol in self.settings.triggers if symbol in content), None)
        for regex in self._regexes:
            found = regex.search(content)
            if found:
                return found.group(0)
        return None

File: rust-agent/crates/core/src/trigger.rs
use regex::Regex;

use crate::{
    config::{ListenConfig, MatchMode},
    models::IncomingMessage,
};

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct TriggerMatch {
    pub room_id: String,
    pub trigger_symbol: String,
    pub trigger_content: String,
}

#[derive(Debug, Clone)]
pub struct TriggerMatcher {
    config: ListenConfig,
    regexes: Vec<Regex>,
}

impl TriggerMatcher {
    pub fn new(config: ListenConfig) -> Result<Self, regex::Error> {
        let regexes = if config.match_mode == MatchMode::Regex {
            config
                .triggers
                .iter()
                .map(|pattern| Regex::new(pattern))
                .collect::<Result<Vec<_>, _>>()?
        } else {
            Vec::new()
        };

        Ok(Self { config, regexes })
    }

    pub fn match_message(&self, msg: &IncomingMessage) -> Option<TriggerMatch> {
        if self.config.ignore_self && msg.is_self {
            return None;
        }

        let room_allowed = self.config.whitelist_rooms.is_empty()
            || self.config.whitelist_rooms.iter().any(|room| {
                room == &msg.room_id
                    || msg
                        .room_name
                        .as_ref()
                        .is_some_and(|name| name.contains(room))
            });
        if !room_allowed {
            return None;
        }

        if self
            .config
            .blacklist_users
            .iter()
            .any(|user| user == &msg.sender_id)
        {
            return None;
        }

        if !self
            .config
            .content_types
            .iter()
            .any(|kind| kind == &msg.msg_type)
        {
            return None;
        }

        let trigger_symbol = match self.config.match_mode {
            MatchMode::Prefix => self
                .config
                .triggers
                .iter()
                .find(|trigger| msg.content.starts_with(trigger.as_str())),
            MatchMode::Contains => self
                .config
                .triggers
                .iter()
                .find(|trigger| msg.content.contains(trigger.as_str())),
            MatchMode::Regex => self
                .regexes
                .iter()
                .zip(self.config.triggers.iter())
                .find_map(|(regex, trigger)| regex.is_match(&msg.content).then_some(trigger)),
        }?;

        Some(TriggerMatch {
            room_id: msg.room_id.clone(),
            trigger_symbol: trigger_symbol.clone(),
            trigger_content: msg.content.clone(),
        })
    }
}

#[cfg(test)]
mod tests {
    use chrono::{TimeZone, Utc};

    use super::*;

    fn config() -> ListenConfig {
        ListenConfig {
            triggers: vec!["/总结".into(), "@".into()],
            match_mode: MatchMode::Prefix,
            whitelist_rooms: vec!["room@chatroom".into()],
            blacklist_users: vec!["wxid_blocked".into()],
            content_types: vec!["text".into()],
            ignore_self: true,
        }
    }

    fn msg(content: &str) -> IncomingMessage {
        IncomingMessage {
            room_id: "room@chatroom".into(),
            room_name: Some("测试群".into()),
            sender_id: "wxid_user".into(),
            sender_name: Some("Alice".into()),
            content: content.into(),
            msg_type: "text".into(),
            timestamp: Utc.timestamp_opt(1_716_464_700, 0).unwrap(),
            is_self: false,
        }
    }

    #[test]
    fn matches_prefix_trigger() {
        let matcher = TriggerMatcher::new(config()).unwrap();
        let matched = matcher.match_message(&msg("/总结 刚才说了什么")).unwrap();
        assert_eq!(matched.trigger_symbol, "/总结");
    }

    #[test]
    fn ignores_non_whitelisted_room() {
        let matcher = TriggerMatcher::new(config()).unwrap();
        let mut message = msg("/总结");
        message.room_id = "other@chatroom".into();
        assert!(matcher.match_message(&message).is_none());
    }

    #[test]
    fn ignores_self_messages() {
        let matcher = TriggerMatcher::new(config()).unwrap();
        let mut message = msg("/总结");
        message.is_self = true;
        assert!(matcher.match_message(&message).is_none());
    }
}
